torch/_higher_order_ops/scan.py

```diff
@@ -0,0 +1,349 @@
+        ), "If init leaves are provided, they must be a Tensor"
+    else:
+        shape = leaves_input[0].shape
+        ndim = len(shape)
+        dim = utils.canonicalize_dim(ndim, dim)
+        num_el = shape[dim]
+        output_spec = spec_input
+
+        # If no init is provided, take the first time step of input as the init
+        # and crop it off the original input
+        leaves_init = [aten.slice(elem, dim, 0, 1, 1) for elem in leaves_input]
+        if num_el > 1:
+            leaves_input = [aten.slice(elem, dim, 1, None, 1) for elem in leaves_input]
+        else:
+            leaves_input = []
+
+    if len(leaves_input) > 0:
+        assert all(
+            isinstance(x, torch.Tensor) for x in leaves_input
+        ), "If input leaves are provided, they must be a Tensor"
+
+        assert all(
+            x.shape[dim] > 0 for x in leaves_input
+        ), "If input leaves are provided, the scan dimension must be > 0"
+
+        out = combine_fn(
+            pytree.tree_unflatten(
+                [aten.slice(elem, dim, 0, 1, 1) for elem in leaves_input], output_spec
+            ),
+            pytree.tree_unflatten(
+                [aten.slice(elem, dim, 0, 1, 1) for elem in leaves_input], output_spec
+            ),
+        )
+        out_leaves, tree_out = pytree.tree_flatten(out)
+        assert len(leaves_input) == len(
+            out_leaves
+        ), "The number of leaves of the pytree of the output of the operator needs to match the lenght of the pytree of the input"
+        for in_l, out_l in zip(leaves_init, out_leaves):
+            assert (
+                in_l.shape == out_l.shape
+            ), "The pytree of the output of the operator needs to match the pytree of the init"
+
+    # Add the init back to the result_flat as the first element
+    if len(leaves_input) > 0:
+        combine_fn = functools.partial(
+            wrap_combine_fn_flat,
+            combine_fn=combine_fn,
+            spec=output_spec,
+            num_leaves=len(leaves_input),
+        )
+
+        result_flat = scan_op(combine_fn, leaves_input, leaves_init, dim)
+
+        if reverse:
+            result_flat = [torch.flip(elem, [dim]) for elem in result_flat]
+
+        return pytree.tree_unflatten(result_flat, output_spec)
+    else:
+        return pytree.tree_unflatten(leaves_init, output_spec)
+
+
+class ScanOp(HigherOrderOperator):
+    def __init__(self):
+        super().__init__("scan")
+
+    def __call__(self, combine_fn, input, init, dim):
+        return super().__call__(combine_fn, input, init, dim)
+
+
+scan_op = ScanOp()
+
+
+def generic_scan(operator, elems_flat, elems_init_flat, dim=0):
+    def combine(a, b, dim):
+        return torch.concatenate([a, b], dim=dim)
+
+    cmb = functools.partial(combine, dim=dim)
+
+    def _scan(elems, elems_init):
+        """Perform scan on `elems`."""
+        num_elems = elems[0].shape[dim]
+
+        ind = 0
+        xs = elems_init
```

Comment:
The naming here is confusing to me lol. We should just rename xs -> inits because the xs in jax's scan actually means elems.

Reply:
I tried to use a better naming now. In particular, I used `input` (to be consistent with associative_scan) and `init` (to be more similar to jax.lax.scan).